[PyTorch][Edge] Adding tests for lite quantized models

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60226 [PyTorch][Edge] Adding tests for lite quantized models**

# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

Commit:
[PyTorch][Edge] Adding tests for lite quantized models

# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

[ghstack-poisoned]

iseeyuan commented on test/mobile/test_lite_script_module.py:
```diff
@@ -1,12 +1,26 @@
+)
+from torch.quantization import (
+    quantize,
+    # prepare,
```
@supriyar and @jerryzh168 , please confirm that if quantize is tested we don't need to use prepare + convert, because they are already used inside quantize.

supriyar replied:
Yes, this should be fine. Do we want to test FX APIs as well?

pavithranrao replied:
@iseeyuan we need to check tests from https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py

iseeyuan commented on test/mobile/test_lite_script_module.py:
```diff
@@ -431,5 +445,98 @@ def forward(self, val: int, x, y, w):
+        with override_quantized_engine(qengine):
+            qconfig = torch.quantization.get_default_qconfig(qengine)
+            model = model_class(**kwargs)
+            # model.qconfig = qconfig
+            # model = prepare(model)
+
+            # test_only_eval_fn(model, self.calib_data)
+            # model = convert(model)
```
Remove?

iseeyuan commented on test/mobile/test_lite_script_module.py:
```diff
@@ -431,5 +445,98 @@ def forward(self, val: int, x, y, w):
+            script_module = torch.jit.script(model)
+            script_module_result = script_module(input)
+
+            buffer = io.BytesIO(script_module._save_to_buffer_for_lite_interpreter())
+            buffer.seek(0)
+            mobile_module = _load_for_lite_interpreter(buffer)
+
+            mobile_module_result = mobile_module(input)
+
+            torch.testing.assert_allclose(script_module_result, mobile_module_result)
+            mobile_module_forward_result = mobile_module.forward(input)
+            torch.testing.assert_allclose(script_module_result, mobile_module_forward_result)
+
+            mobile_module_run_method_result = mobile_module.run_method("forward", input)
+            torch.testing.assert_allclose(script_module_result, mobile_module_run_method_result)
```
One issue we had is flaky: the test crashes very 2-3 times. Could you put these part in a loop and run it around 5 times? If there's signal of CI performance we could reduce the number of times.

iseeyuan reviewed (commented):
Looks good to me in general. Added some inline comments.

@supriyar and @jerryzh168 , could you take a look and let us know if there are more tests (features or models) we should include?

Commit:
Update base for Update on "[PyTorch][Edge] Adding tests for lite quantized models"


# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Adding tests for lite quantized models"


# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

[ghstack-poisoned]

supriyar commented on test/mobile/test_lite_script_module.py:
```diff
@@ -431,5 +445,109 @@ def forward(self, val: int, x, y, w):
+            # model = convert(model)
+            model = quantize(model, test_only_eval_fn, [self.calib_data])
+
+            script_module = torch.jit.script(model)
+            script_module_result = script_module(input)
```
Do we want to do jit.script save-load also to be consistent?

pavithranrao replied:
I think `save` and `load` functions would be tested for consistency in pytorch-jit test bench. If you feel we should save and load script models, we can add them too. cc: @iseeyuan

iseeyuan replied:
I think save and load are covered by _save_for_lite and _load_for_lite. Basically, _save_for_lite would call save, plus the mobile part.

supriyar commented on test/mobile/test_lite_script_module.py:
```diff
@@ -431,5 +445,109 @@ def forward(self, val: int, x, y, w):
+
+    def test_single_layer(self):
+        input = torch.rand(2, 5, dtype=torch.float)
+        self._compare_script_and_mobile(model_class=AnnotatedSingleLayerLinearModel, input=input, qengine="qnnpack")
```
Is the `qengine` argument being used?

pavithranrao replied:
Some modules need `qengine`, so `qengine` is provided as `kwargs` in `_compare_script_and_mobile` like this:

```
model = model_class(**kwargs)
```

supriyar commented on test/mobile/test_lite_script_module.py:
```diff
@@ -431,5 +445,109 @@ def forward(self, val: int, x, y, w):
+        input = torch.rand(2, 5, dtype=torch.float)
+        self._compare_script_and_mobile(model_class=AnnotatedNestedModel, input=input, qengine="qnnpack")
+
+    def test_quantization_example(self):
```
Why aren't we using `_compare_script_and_mobile` for this test as well?

pavithranrao replied:
you are right, I will try and split `_compare_script_and_mobile` into two functions. One to create a module object and the other to compare output tensors (script and lite).

Commit:
Update base for Update on "[PyTorch][Edge] Adding tests for lite quantized models"


# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch][Edge] Adding tests for lite quantized models"


# Context
Read this posts for details about why we need a test bench for quantized lite modules
https://fb.workplace.com/groups/2322282031156145/permalink/4289792691071726/

# This Diff
Adds test cases for Quantized Lite modules

Differential Revision: [D29212232](https://our.internmc.facebook.com/intern/diff/D29212232/)

[ghstack-poisoned]

iseeyuan reviewed (approved):
LGTM. 

Please add coverage on FX flow as @supriyar suggested: "https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py"

FX coverage can be made in a different PR.
